kallsyms/kallsyms.go

```diff
@@ -601,20 +605,104 @@ func (s *Symbolizer) pollKobjectClient(kobjectClient *kobject.Client) {
 			// Notify worker thread without blocking
 			select {
 			case s.reloadModules <- true:
+			case <-ctx.Done():
+				return
 			default:
 			}
 		}
 	}
 }
 
+// bpfKsymsTrigger attaches a minimal eBPF program to bpf events, that update
+// kallsyms, to notify user space about changes.
+func (s *Symbolizer) bpfKsymsTrigger(ctx context.Context) {
+	events, err := ebpf.NewMap(&ebpf.MapSpec{
+		Type: ebpf.PerfEventArray,
+		Name: "ksym_update_trigger",
+	})
+	if err != nil {
+		log.Errorf("Failed creating perf event array: %s", err)
+		return
+	}
+	defer events.Close()
+
+	rd, err := perf.NewReader(events, os.Getpagesize())
+	if err != nil {
+		log.Errorf("Failed creating event reader: %s", err)
+		return
+	}
+	defer rd.Close()
+
+	var progSpec = &ebpf.ProgramSpec{
+		Name:    "ksym_update_trigger",
+		Type:    ebpf.Kprobe,
+		License: "GPL",
+	}
+
+	// Minimal eBPF program that writes the static value '1' to the perf event
+	// array on each event. The arbitrary value of '1' is used just to notify
+	// the user space about changes to kallsyms.
+	progSpec.Instructions = asm.Instructions{
+		// store the integer 1 at FP[-8]
+		asm.Mov.Imm(asm.R2, 1),
+		asm.StoreMem(asm.RFP, -8, asm.R2, asm.Word),
+
+		// load registers with arguments for call of FnPerfEventOutput
+		asm.LoadMapPtr(asm.R2, events.FD()), // file descriptor of the perf event array
+		asm.LoadImm(asm.R3, 0xffffffff, asm.DWord),
+		asm.Mov.Reg(asm.R4, asm.RFP),
+		asm.Add.Imm(asm.R4, -8),
+		asm.Mov.Imm(asm.R5, 4),
+
+		// call FnPerfEventOutput, an eBPF kernel helper
+		asm.FnPerfEventOutput.Call(),
+
+		// set exit code to 0
+		asm.Mov.Imm(asm.R0, 0),
+		asm.Return(),
+	}
+
+	prog, err := ebpf.NewProgram(progSpec)
+	if err != nil {
+		log.Errorf("Failed creating ebpf program: %s", err)
+		return
+	}
+	defer prog.Close()
+
+	kp, err := link.Kprobe("bpf_ksym_add", prog, nil)
+	if err != nil {
+		log.Errorf("Failed opening kprobe: %s", err)
+		return
+	}
+	defer kp.Close()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			if _, err := rd.Read(); err != nil {
+				if errors.Is(err, perf.ErrClosed) {
+					return
+				}
+				log.Errorf("Failed to handle notification on bpf_ksym_add: %v", err)
+				continue
+			}
+			s.reloadModules <- true
```

Comment:
Can we switch this to async send?


```suggestion
			select {
			case s.reloadModules <- true:
			default:
			}
```